lib/Optimizer/Transforms/ApplyOpSpecialization.cpp

```diff
@@ -75,20 +76,67 @@ struct ApplyOpAnalysis {
 private:
   void performAnalysis(Operation *op) {
     op->walk([&](quake::ApplyOp apply) {
+      if (constProp) {
+        // If some of the arguments in getArgs() are constants, then materialize
+        // those constants in a clone of the variant. The specialized variant
+        // will then be able to perform better constant propagation even if not
+        // inlined.
+        auto calleeName = apply.getCallee()->getRootReference().str();
+        if (func::FuncOp genericFunc =
+                module.lookupSymbol<func::FuncOp>(calleeName)) {
+          SmallVector<Value> newArgs;
+          newArgs.append(apply.getArgs().begin(), apply.getArgs().end());
+          bool arithConsts = false;
+          IRMapping mapper;
+          SmallVector<arith::ConstantOp> moveConsts;
+          for (auto [idx, v] : llvm::enumerate(newArgs)) {
+            if (auto c = v.getDefiningOp<arith::ConstantOp>()) {
+              auto newConst = c.clone();
+              moveConsts.push_back(newConst);
+              mapper.map(genericFunc.getArgument(idx), newConst);
+              LLVM_DEBUG(llvm::dbgs() << "apply has constant arguments.\n");
+              arithConsts = true;
+            }
+          }
+
+          if (arithConsts) {
```

Comment:
I wonder if we can get rid of `arithConsts` and use `!moveConsts.empty()` instead.